File: 4.8-bc/py/stage-1-flat/app/service/order.py
from __future__ import annotations

from datetime import UTC, datetime
from uuid import UUID, uuid4

from app.model.order import Order
from app.repository.order import OrderRepo


class OrderService:
    def __init__(self, repo: OrderRepo) -> None:
        self._repo = repo

    async def place(self, user_id: UUID, total_cents: int) -> Order:
        o = Order(
            id=uuid4(),
            user_id=user_id,
            total_cents=total_cents,
            status="pending",
            created_at=datetime.now(UTC),
        )
        await self._repo.create(o)
        return o

File: 4.8-bc/go/stage-1-flat/app/service/order.go
package service

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-1-flat/app/model"
)

type OrderRepo interface {
	Create(ctx context.Context, o model.Order) error
}

type OrderService struct {
	repo OrderRepo
}

func NewOrderService(r OrderRepo) *OrderService {
	return &OrderService{repo: r}
}

func (s *OrderService) Place(ctx context.Context, userID uuid.UUID, totalCents int64) (*model.Order, error) {
	o := model.Order{
		ID:         uuid.New(),
		UserID:     userID,
		TotalCents: totalCents,
		Status:     "pending",
		CreatedAt:  time.Now().UTC(),
	}
	if err := s.repo.Create(ctx, o); err != nil {
		return nil, err
	}
	return &o, nil
}
